service/matching/db_task_manager.go

```diff
@@ -0,0 +1,258 @@
+			taskQueueKey,
+			store,
+			taskOwnership.getAckedTaskID(),
+			logger,
+		),
+		dispatchTaskFn: dispatchTaskFn,
+		finishTaskFn:   finishTaskFn,
+		logger:         logger,
+
+		shutdownChan: make(chan struct{}),
+		dispatchChan: make(chan struct{}, 1),
+	}, nil
+}
+
+func (d *dbTaskManager) Start() {
+	if !atomic.CompareAndSwapInt32(
+		&d.status,
+		common.DaemonStatusInitialized,
+		common.DaemonStatusStarted,
+	) {
+		return
+	}
+
+	d.SignalDispatch()
+	go d.readerEventLoop()
+	go d.writerEventLoop()
+}
+
+func (d *dbTaskManager) Stop() {
+	if !atomic.CompareAndSwapInt32(
+		&d.status,
+		common.DaemonStatusStarted,
+		common.DaemonStatusStopped,
+	) {
+		return
+	}
+
+	close(d.shutdownChan)
+}
+
+func (d *dbTaskManager) SignalDispatch() {
+	select {
+	case d.dispatchChan <- struct{}{}:
+	default: // channel already has an event, don't block
+	}
+}
+
+func (d *dbTaskManager) isStopped() bool {
+	return atomic.LoadInt32(&d.status) == common.DaemonStatusStopped
+}
+
+func (d *dbTaskManager) writerEventLoop() {
+	updateQueueTicker := time.NewTicker(dbTaskUpdateQueueInterval)
+	defer updateQueueTicker.Stop()
+	// TODO we should impl a more efficient method to
+	//  buffer & wait for max duration
+	//  right now simply just flush every dbTaskFlushInterval
+	flushTicker := time.NewTicker(dbTaskFlushInterval)
+	defer flushTicker.Stop()
+
+	for {
+		if d.isStopped() {
+			return
+		}
+
+		select {
+		case <-d.shutdownChan:
+			return
+		case <-d.taskOwnership.getShutdownChan():
+			d.Stop()
+
+		case <-updateQueueTicker.C:
+			d.persistTaskQueue()
+		case <-flushTicker.C:
+			d.taskFlusher.flushTasks()
+			d.SignalDispatch()
+		case <-d.taskFlusher.notifyFlushChan():
+			d.taskFlusher.flushTasks()
+			d.SignalDispatch()
+		}
+	}
+}
+
+func (d *dbTaskManager) readerEventLoop() {
+	updateAckTicker := time.NewTicker(dbTaskUpdateAckInterval)
+	defer updateAckTicker.Stop()
+
+	for {
+		if d.isStopped() {
+			return
+		}
+
+		select {
+		case <-d.shutdownChan:
+			return
+		case <-d.taskOwnership.getShutdownChan():
+			d.Stop()
+
+		case <-updateAckTicker.C:
+			d.updateAckTaskID()
+		case <-d.dispatchChan:
+			d.readDispatchTask()
+		}
+	}
+}
+
+func (d *dbTaskManager) writeAppendTask(
+	task *persistencespb.TaskInfo,
+) future.Future {
+	return d.taskFlusher.appendTask(task)
+}
+
+func (d *dbTaskManager) readDispatchTask() {
+	iter := d.taskTracker.taskIterator(d.taskOwnership.getLastAllocatedTaskID())
+	for iter.HasNext() {
+		item, err := iter.Next()
+		if err != nil {
+			d.logger.Error("dbTaskManager encountered error when fetching tasks", tag.Error(err))
+			d.SignalDispatch()
+			return
+		}
+
+		task := item.(*persistencespb.AllocatedTaskInfo)
+		d.mustDispatch(task)
+	}
+}
+
+func (d *dbTaskManager) mustDispatch(
+	task *persistencespb.AllocatedTaskInfo,
+) {
+	for !d.isStopped() {
+		if taskqueue.IsTaskExpired(task) {
+			d.taskTracker.ackTask(task.TaskId)
+			return
+		}
+
+		err := d.dispatchTaskFn(newInternalTask(
```

Comment:
if anything goes wrong, we will fall into a busy spin here?

Reply:
there is a `for !d.isStopped() {` at the beginning
`d.finishTaskFn,` will try to move this function from head of queue to end of queue (which i believe is wrong, but too many changes already)
plus, this (at least right now) only mimics the current behavior

Reply:
I think this needs to be addressed. We should avoid retry in busy loop. It can be a TODO with follow up PR.

Reply:
there will be following PRs
quite lots of places to optimize & simplify
but implementation (rewrite) should be done one step at a time
PRs already sent out / merged are focused on db task read / write
later there should be sync match related PRs & matching logic governing sync / async match policy